### PR TITLE
A limit on the number of materialized views attached to a table.

### LINKED ISSUE #56428: A limit on the number of materialized views attached to a table.
A limit can be configurable by a query-level setting (the setting could be constrained then).

It can be set to 100 by default.

### PR DESCRIPTION
**Pull request by zhongyuankai**
Closes #56428
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add settings `max_materialized_views_size_for_table` to limit the number of materialized views attached to a table.

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -847,6 +847,7 @@ class IColumn;
     M(Int64, ignore_cold_parts_seconds, 0, "Only available in ClickHouse Cloud", 0) \
     M(Int64, prefer_warmed_unmerged_parts_seconds, 0, "Only available in ClickHouse Cloud", 0) \
     M(Bool, enable_order_by_all, true, "Enable sorting expression ORDER BY ALL.", 0)\
+    M(UInt64, max_materialized_views_size_for_table, 100, "A limit on the number of materialized views attached to a table.", 0)\
```
Comment (nickitat) [inline src/Core/Settings.h:850]: size of smth and amount of smth are different things

File: src/Storages/StorageMaterializedView.cpp [hunk 2]
```diff
@@ -83,6 +84,11 @@ StorageMaterializedView::StorageMaterializedView(
         throw Exception(ErrorCodes::QUERY_IS_NOT_SUPPORTED_IN_MATERIALIZED_VIEW, "UNION is not supported for MATERIALIZED VIEW");
 
     auto select = SelectQueryDescription::getSelectQueryFromASTForMatView(query.select->clone(), local_context);
+    auto select_table_dependent_views = DatabaseCatalog::instance().getDependentViews(select.select_table_id);
+    if (select_table_dependent_views.size() >= local_context->getSettingsRef().max_materialized_views_size_for_table)
```
Comment (nickitat) [inline src/Storages/StorageMaterializedView.cpp:88]:
```suggestion
    if (max_materialized_views_size_for_table && select_table_dependent_views.size() >= max_materialized_views_size_for_table)
```

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit ff8d164e4afc3a78ca5fc71c8aa85744d04a644d with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58068/ff8d164e4afc3a78ca5fc71c8aa85744d04a644d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7277626867">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58068/ff8d164e4afc3a78ca5fc71c8aa85744d04a644d/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by bharatnc** [comment]
Hi, @nickitat can you help add the label to trigger the tests?

File: docs/en/operations/settings/settings.md [hunk 3]
```diff
@@ -2992,6 +2992,12 @@ Default value: 268435456.
 
 - [min_insert_block_size_bytes](#min-insert-block-size-bytes)
 
+## max_materialized_views_account_for_table {#max_materialized_views_account_for_table}
```
Comment (nickitat) [inline docs/en/operations/settings/settings.md:2995]:
```suggestion
## max_materialized_views_count_for_table {#max_materialized_views_count_for_table}
```

File: docs/en/operations/settings/settings.md [hunk 4]
```diff
@@ -2992,6 +2992,12 @@ Default value: 268435456.
...
+
+A limit on the number of materialized views attached to a table.
```
Comment (nickitat) [inline docs/en/operations/settings/settings.md:2997]: iirc, it is possible to create view over another view. let's clarify if we account those outer views or not

File: docs/en/operations/settings/settings.md [hunk 5]
```diff
@@ -2992,6 +2992,12 @@ Default value: 268435456.
...
+
+Default value: `100`.
```
Comment (nickitat) [inline docs/en/operations/settings/settings.md:2999]:
```suggestion
Default value: `0`.
```

File: tests/queries/0_stateless/02947_limit_materialized_view_size.sql [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+CREATE MATERIALIZED VIEW 02947_tb_view2
+ENGINE = MergeTree
+ORDER BY c
+AS SELECT * FROM 02947_tb;  -- { serverError TOO_MANY_MATERIALIZED_VIEWS }
+
+CREATE MATERIALIZED VIEW 02947_tb_view2
+ENGINE = MergeTree
+ORDER BY c
+AS SELECT * FROM 02947_tb
+SETTINGS max_materialized_views_size_for_table = 2;
```
Comment (nickitat) [inline tests/queries/0_stateless/02947_limit_materialized_view_size.sql:30]: it looks a little bit strange as a query-level setting. shouldn't it be a server setting instead?

File: docs/en/operations/settings/settings.md (hunk 4 above)
Comment (zhongyuankai) [reply docs/en/operations/settings/settings.md]: updated

File: tests/queries/0_stateless/02947_limit_materialized_view_size.sql (hunk 6 above)
Comment (zhongyuankai) [reply tests/queries/0_stateless/02947_limit_materialized_view_size.sql]: Yes, I have changed to service settings.

**Comment by nickitat** [comment]
> @nickitat Do I need to change to integration testing?

yes, I think it would be the best solution

**Review by nickitat — APPROVED** [review]

**Comment by zhongyuankai** [comment]
@nickitat Hi, can it be merged into master?

**Pull request merged**